### ISSUE TITLE
Create a new microservice for attachments

### ISSUE THREAD

**Issue by maxceem** [appirio-tech/connect-app#3474]
This is not actually a task for Connect App, but we don't have a separate repository for a new microservice yet.

We are thinking of having a separate service for managing attachments and then migrating the data from project service and message service to that attachment service and then use that service to provide common features like filtering and sorting on all attachments (files and links) of the projects.

There are multiple github issues related to this overall work Connect App:
- https://github.com/appirio-tech/connect-app/issues/3466
- https://github.com/appirio-tech/connect-app/issues/3465
- https://github.com/appirio-tech/connect-app/issues/3464
- https://github.com/appirio-tech/connect-app/issues/3463
- https://github.com/appirio-tech/connect-app/issues/3462

I guess we have to create something like we have for topics in Message Service with `reference` and `referenceId`. In our case `reference` could be `project` for projects, also, we might have `reference` as `phase` or work for the attachments inside works if eventually, we add the ability to upload files to `works`.

**Comment by maxceem** [comment]
@vikasrohit I've noticed, that most new Topcoder microservices use DynamoDB as a Database. Should we do the same or should we use PosgreSQL as we do for Project Service?

- `resources-api` https://github.com/topcoder-platform/resources-api/blob/develop/src/models/index.js
- `challenge-api` https://github.com/topcoder-platform/challenge-api/blob/develop/src/models/index.js
- `submissions-api` https://github.com/topcoder-platform/submissions-api/blob/develop/src/common/dbhelper.js
- `lookups-api` - https://github.com/topcoder-platform/lookups-api/blob/develop/src/models/index.js

**Comment by vikasrohit** [comment]
@maxceem here are my answers:

1. postgres
2. yes
3. I am not sure in this case as I don't think we would like to keep the actual files for the soft deleted attachments and attachments does not make sense without their actual file. Lets discuss this further.
4. For now we can assume this as `/v5/attachments`.

I have to discuss this new microservice with broader team before we can actually launch it. I would try to get nods for it by this weekend, however, I think most probably it would spill over to next week.

**Comment by maxceem** [comment]
@vikasrohit I've almost ready with the spec for the initial code for the attachment service. There is one thing I wanted to discuss.

In general, we want to implement permissions for attachments, the same as we do in Message Service.
- If we request the `attachments` for the `project`, we load that `project` with the same user token, and  if a `project` is accessible, it means that we also can access the `attachments` (ignore the requirement to support "shared with" for now)
- for `posts` or `topics` same, if we want to get `attachments` for some `post` or `topic`, we load the `topic`, and if we can access it, then we also can get the `attachments`

Now we have the requirement to filter across all `attachment` for the whole `project` which also includes attachments for all the `topics` and `posts` in this issue https://github.com/appirio-tech/connect-app/issues/3464. So we have to implement an endpoint to get all the `attachments` for the project including attachments for posts inside topics. Creating such an endpoint with such a filter is clear.

Though there is a tricky part with permissions. We have `private` topics that cannot be seen by `customer` users. So for **each** of the attachment, we should load respective topic to check if the user has permission to access it or no.

So when we would call such endpoint to get all the attachemnts for the project including attachments inside posts, we have to internally make as many calls to Message Service as we have topics.

I don't know how many topics per project we have at the moment in average, do you think that would be ok for now or we have to think for some optimizations from the beginning for this?

**Comment by vikasrohit** [comment]
I guess, we won't need the separate call for validating the private topics. We would be calling the `GET /v5/topics/1234` as reference look up call, we can parse the private flag at this moment itself via `accessCondition` to validate if the user has access to the attachments of the topic.

What I am worried now, is if we have to make separate (at least one for topics and one for project) calls to fetch attachments of a project, we can not have the common filtering and sorting feature to the UI.

**Comment by maxceem** [comment]
> What I am worried now, is if we have to make separate (at least one for topics and one for project) calls to fetch attachments of a project, we can not have the common filtering and sorting feature to the UI.

I think there is quite a straightforward solution. The general idea instead of having **one** reference per attachment we should have **multiple** `references`, so after we can filter by them, for example:

### Model

#### Attachment object for a `post` has 3 references

```js
{
  id: 1,
  references: [
   { entity: 'project', enityId: 1234 },
   { entity: 'topic', enityId: 12 },
   { entity: 'post', enityId: 123 }
  ]
}
```

#### Attachment object for a `project` has 1 reference

```js
{
  id: 2,
  references: [
   { entity: 'project', enityId: 1234 }
  ]
}
```

#### Attachment object for a `work` has 2 references

```js
{
  id: 3,
  references: [
   { entity: 'project', enityId: 1234 },
   { entity: 'work', enityId: 234 }
  ]
}
```

### Queries

#### All attachments for a `project` including `post` and `work` attachments

- `GET /attachments?includeReferences=entity:project,entityId:1234`
- DB/ES query logic: `SELECT attachments WHERE references CONTAINS { entity: "project", entityId: 1234 }`

#### Only `project` attachments

If we want to get `attachments` only for `project` without such for `posts` and `works`, we can call:

- `GET /attachments?onlyReferences=entity:project,entityId:1234`
- DB/ES query logic: `SELECT attachments WHERE references CONTAINS **ONLY** { entity: "project", entityId: 1234 }`

**The difference between 2 queries above in using `includeReferences` or `onlyReferences` query param**

#### All `post` attachments inside the project

- `GET /attachments?onlyReferences=entity:project,enityId:1234&onlyReferences=entity:topic&onlyReferences=entity:post`
- DB/ES query logic: `SELECT attachments WHERE references CONTAINS **ONLY** ( { entity: "project", entityId: 1234 } AND { entity: "topic" } AND { entity: "post" } )`

#### All `work` attachments inside the project

- `GET /attachments?onlyReferences=entity:project,enityId:1234&onlyReferences=entity:work`
- DB/ES query logic: `SELECT attachments WHERE references CONTAINS **ONLY** ( { entity: "project", entityId: 1234 } AND { entity: "work" } )`

### Problems to solve

The most critical part of the idea above is that we can create real PostreSQL queries and ElasticSearch queries for such cases. We might adjust data model and API query params to achieve faster queries.
- First of all queries to ElasticSearch should be very fast as in 99.99% cases we should get data from ES. I guess our aim time`< 0.5 seconds for 10 Million records`
- Queries to DB are not required to be super fast, but it shouldn't keep DB busy for too long. I guess our aim could be `< 5 seconds for 10 Million records`

Here is a quick POC for DB https://dbfiddle.uk/?rdbms=postgres_10&fiddle=303ff75d14b866823189b809d1c165ad

**Comment by vikasrohit** [comment]
Thanks @maxceem for alternative approach and in fact we thought of the same yesterday, however, the blocker where we are stuck is that how would we validate the access to such attachments. I mean, when we are querying attachments for `GET /attachments?includeReferences=entity:project,entityId:1234` call, it would return the attachments for admin posts as well, unless we call and verify every reference look up endpoint (i.e. for topic attachments, we have to call `/v5/topics/{id}` for every topic which has attachment) for all attachments.

**Comment by maxceem** [comment]
Yeah, here is where we have to come with some solution. So if use the current approach with reference endpoint for `GET /attachments?includeReferences=entity:project,entityId:1234` we would call:
- 1 `GET /project` to verify access to the project and works
- 100 `GET /topic/{id}` to verify access to each topic (in case project has 100 topics)

This could lead to a lot of requests per **each referenced item**.

I didn't think this through completely, though I guess the solution might be to replace endpoints in `ReferenceLookup` table with requests to get **lists of items**, instead of items individually, something like this:

```js
// ReferenceLookups
[
  { 
     references: ["project"], 
     listEndpoint: "https://api.topcoder-dev.com/v5/projects?id={projectId}", 
     itemSelector: "{ id: '{projectId}' }" 
  },
  { // for works we can use the same
     references: ["project", "work"],  
     listEndpoint: "https://api.topcoder-dev.com/v5/projects?id={projectId}", 
     itemSelector: "{ id: '{projectId}' }" 
  },
  { // or imagine that "customers" cannot access "draft" works, 
    // then for works we have to check permissions in a differene way than for projects
     references: ["project", "work"]  
     // we actually don't have such endpoint at the moment for works, so we would have to create it
     listEndpoint: "https://api.topcoder-dev.com/v5/projects/{projectId}/works", 
     itemSelector: "{ id: '{workId}' }" 
  }, 
  { 
     references: ["project", "topic", "post"], 
     listEndpoint: "https://api.topcoder-dev.com/v5/topics?reference=project&referenceId={projectId}", 
     itemSelector: "{ id: '{topicId}' }" 
  }
]
```

So imagine, that we want to get all attachments for the project and calling `GET /attachments?includeReferences=entity:project,entityId:1234`.

We got from the DB next attachments and we have to check permissions for them:
```js
[
   // 2 for project
   { id: 1, references: [{ entity: "project", enitityId: 1234 }] },
   { id: 2, references: [{ entity: "project", enitityId: 1234 }] },
   // 2 for works
   { id: 3, references: [{ entity: "project", enitityId: 1234 }, { entity: "work", entityId: 123 }] },
   { id: 4, references: [{ entity: "project", enitityId: 1234 }, { entity: "work", entityId: 456 }] },
   // 2 for topics
   { id: 5, references: [{ entity: "project", enitityId: 1234 }, { entity: "topic", entityId: 12 }, { entity: "post", entityId: 1111 }] },
   { id: 6, references: [{ entity: "project", enitityId: 1234 }, { entity: "topic", entityId: 45 }, { entity: "post", entityId: 2222 }] },
]
```

- First, we extract from the list of attachments we've got, that we have 3 cases of `references`:
   - `references: ["project"]`
   - `references: ["project", "work"]`
   - `references: ["project", "topic", "post"]`
- So we have to call only 3 endpoints to get referenced objects and get permissions.
- Imagine we've got a list of topics by calling `https://api.topcoder-dev.com/v5/topics?reference=project&referenceId=1234`:
   ```js
    topics = [
     { id: 12 }
    ]
   ```
    And now we have to check permissions for each attachments that have `references: ["project", "topic", "post"]`:
    - for ` { id: 5, references: [{ entity: "project", enitityId: 1234 }, { entity: "topic", entityId: 12 }, { entity: "post", entityId: 1111 }] }` we have `itemSelector = { id: 12}`, so we search for the topic by `_.find(topics, itemSelector)` and we found the topic. So we have the access. Additionallly, we might evaluate `accessCondition` by expression evaluator for the topic.
  - for ` { id: 6, references: [{ entity: "project", enitityId: 1234 }, { entity: "topic", entityId: 45 }, { entity: "post", entityId: 2222 }] }` we have `itemSelector = { id: 45}`, so we search for the topic by `_.find(topics, itemSelector)` but we cannot find topic, so we don't have access.

### What if another service (for example `challenge-service`) also has `topics` with `attachments`.

Then attachments would have the `challenge` reference instead of `project`:
```js
[
   // 2 for topics of challenges
   { id: 7, references: [{ entity: "challenge", enitityId: 30303003 }, { entity: "topic", entityId: 12 }, { entity: "post", entityId: 1111 }] },
   { id: 8, references: [{ entity: "challenge", enitityId: 40404004 }, { entity: "topic", entityId: 45 }, { entity: "post", entityId: 2222 }] },
]
```

And we have to create an additional `ReferenceLookup` record:
```js
  { 
     references: ["challenge", "topic", "post"], 
     listEndpoint: "https://api.topcoder-dev.com/v5/topics?reference=challenge&referenceId={challengeId}", 
     itemSelector: "{ id: '{topicId}' }" 
  }
]
```

**Comment by vikasrohit** [comment]
@maxceem are you suggesting that we should update the schema of ReferenceLookup table to have structure like
```
{ 
     references: ["challenge", "topic", "post"], 
     listEndpoint: "https://api.topcoder-dev.com/v5/topics?reference=challenge&referenceId={challengeId}", 
     itemSelector: "{ id: '{topicId}' }" 
  }
```
OR this is just a shortcut way you are using the show the 3 separate records?

**Comment by vikasrohit** [comment]
Also, even if we make only 3 calls to validate the access to the attachments, I guess, we still can not have a seamless paging because we would be filtering the attachments after fetching them from database, right?

**Comment by maxceem** [comment]
> @maxceem are you suggesting that we should update the schema of ReferenceLookup table to have structure like

```js
{ 
     references: ["challenge", "topic", "post"], 
     listEndpoint: "https://api.topcoder-dev.com/v5/topics?reference=challenge&referenceId={challengeId}", 
     itemSelector: "{ id: '{topicId}' }" 
  }

Yes, I suggest updating `ReferenceLookup` table structure like this to support multiple references and `listEndpoints`.

**Comment by maxceem** [comment]
> Also, even if we make only 3 calls to validate the access to the attachments, I guess, we still can not have a seamless paging because we would be filtering the attachments after fetching them from database, right?

Good catch. You are right implementing seamless pagination is not straightforward:
- we either would have to request all the records from DB/ES and after paginate them using JS
- or we can request only one page of data from DB/ES, after filter it by permission, and if we end up having less records than for 1 page, we should request data from DB/ES again. And we should repeat it until we have enough records for one page or there is no more data in DB/ES for our query
- or we can combine these 2 approaches: first request data for one page, but if after permission filter we don't have enough records for one page, we can request all data

Though such logic seems to me quite tricky, feels like not so good idea to create a new service that from start would have such `custom` not straightforward logic. I would think about other solutions.

**Comment by vikasrohit** [comment]
> feels like not so good idea to create a new service that from start would have such custom not straightforward logic. I would think about other solutions.

Agreed. That is why I am not yet fully convinced to create the new service. Lets think more about it and then decide. Lets focus on finishing the features around Assets library targeted in this release 2.5.1 and keep discussing this new API along with.

**Comment by maxceem** [comment]
@vikasrohit as we are not ready for creating a separate service, we've recently decided to update attachments functionality in Project Service to support links and tags inside Project Service.

I'm close to finishing the spec https://docs.google.com/document/d/184pCthxtttLk4fce9Es8Kljo9D45QxUiRUMfbOEN_bk/edit#heading=h.m73b0koll5f6 and would like to confirm a few moments:
1. We should update [ProjectAttachment](https://github.com/topcoder-platform/tc-project-service/blob/develop/src/models/projectAttachment.js#L4) model to support links and tags:
    Add fields:
    - “url” - String, optional - URL for links
    - “type” - String, required, allowed values: “file” and “link” - determines if the attachment is link or file
    - “tags” - JSON, required, default “[]” - JSON array of string tags, example `[“specification”, “design preview”, “billing information”]`

    Update fields:
     - “filePath” - make it optional as only files would have it

2. We have event `connect.notification.project.files.updated` that we consume internally by Project Service. We use it to update `lastActivityAt` of projects. We don't use it in TC Notification Service or in any other place that I'm aware of.

   Now we would need to update `lastActivityAt` for links the same way as for files. As `connect.notification.project.files.updated` is called when any attachment is updated it would be naturally called when any link is updated. The only thing that its name doesn't suite its new role so I propose to rename it to `connect.notification.project.attachments.updated`.

3. We would create a JS script to migrate links from `project.bookmarks` to `ProjectAttachment` model. OK?

4. As we would migrate links from `project.booksmarks` field, would you like to remove this field now? Or would you like to keep it for some time and remove it later? We would also have to remove the corresponding code and unit test regarding `project.booksmarks` field.

**Comment by vikasrohit** [comment]
Yes, I guess for now have to go with changes in existing attachments model. Do you think, we need to do the same changes to the messages service to have seamless integration in future when we try to merge both attachments models into new service?

Answers to the specific questions:
1. Looks good. Just an idea, may be we can reuse the `filePath` field to store the URL as well by renaming it to `path`.
2. Makes sense
3. Yes, JS migration script should work because it would be difficult to have SQL migration script as it involves parsing JSON and assigning some default values to the audit fields.
4. I would like to keep them as is for some time, we can log an issue to track final removal of that field. That reminds me, is it possible for us to have a reverse script(updating bookmarks field from the new table) for backup plan?

**Comment by maxceem** [comment]
> Do you think, we need to do the same changes to the messages service to have seamless integration in future when we try to merge both attachments models into new service?

I don't see the benefits of this. So I suggest keeping as it is for now, until we come with some new requirements or decide on a separate service.

> 1. Looks good. Just an idea, may be we can reuse the filePath field to store the URL as well by renaming it to path.

Yes, it makes sense.

> is it possible for us to have a reverse script(updating bookmarks field from the new table) for backup plan?

I think yes, I would ask to create a backward script as well. So I would also ask to keep the code regarding `bookmarks` workable.

**Comment by vikasrohit** [comment]
> I think yes, I would ask to create a backward script as well. So I would also ask to keep the code regarding bookmarks workable.

And please log a ticket to get rid of this code in near future. May be we can ask in the challenge to make a list of such items when we actually remove it,

**Comment by maxceem** [comment]
> And please log a ticket to get rid of this code in near future. May be we can ask in the challenge to make a list of such items when we actually remove it,

A follow-up ticket is created https://github.com/topcoder-platform/tc-project-service/issues/471
